### DISCUSSION TITLE
Bundle size

### DISCUSSION THREAD

**Discussion by axeldelafosse** [rainbow-me/rainbowkit#699] [General]
Hey!

I installed `@next/bundle-analyzer` in the example app `with-next` and found this:

```
Route (pages)                              Size     First Load JS
┌ ○ /                                      309 B           298 kB
├   /_app                                  0 B             297 kB
└ ○ /404                                   195 B           298 kB
+ First Load JS shared by all              302 kB
  ├ chunks/framework-b8328cdda6d53bb1.js   45.3 kB
  ├ chunks/main-d9f259fa8d709dab.js        30.9 kB
  ├ chunks/pages/_app-c78a6b45312c3139.js  219 kB
  ├ chunks/webpack-31b7295a30eedaa2.js     2.09 kB
  └ css/9d420c3caa707bff.css               4.45 kB
```

300 kB is huge! For comparison, the goal should be less than 14 kB: https://endtimes.dev/why-your-website-should-be-under-14kb-in-size/. Look at the `_app.js` here:

<img width="3008" alt="Screenshot 2022-08-30 at 13 30 54" src="https://user-images.githubusercontent.com/10477267/187426466-641c8714-6788-4505-9335-269f243d8e3e.png">

Most of it comes from importing `ethers`, `rainbowkit` and `wagmi`.

What can we do to improve this?

- Can we improve the tree-shaking?
- Can we lazy load some parts?
- Can we reduce the bundle size of the libraries?

For example, I'm working on a PR for the Showtime app where I'm improving `ethers` tree-shaking by avoiding `import { ethers } from "ethers";` and importing the specific package from `@ethersproject/` and also lazy loading some by using `const Web3Provider = (await import("@ethersproject/providers")).Web3Provider;` for example. Here is the PR: https://github.com/showtime-xyz/showtime-frontend/pull/1459

To paraphrase Vercel: let's make the web faster! If every dapp is slow because of the libs mentioned earlier, this is a pretty bad situation for web3.

Happy to help!

**Comment by peduarte** [comment]
> If every dapp is slow because of the libs mentioned earlier, this is a pretty bad situation for web3.

are you noticing the app being slow specifically because of rainbowkit (inc its deps)? or is this solely from the bundle-analyzer results? for the record i totally agree that we should aim for small bundle sizes and performant applications, just curious on your experience so far.

since we depend heavily on ethers (and so does wagmi), it's hard to shave off that ~140kb bundle size dependency

i like your suggestions and would be open to welcome such a contribution! <3 

cc'ing @markdalgleish here in case this is something he's been thinking about already

**Reply by axeldelafosse** [reply]
Hey Pedro!

I'm currently digging into our Web Vitals Metrics from Next.js on Vercel and we don't have a good score. One of the reason is the huge bundle size. We are using React Native for Web, React Native Reanimated and React Native Gesture Handler so they already account for 300+ kB. There is already some work in progress to improve this. Another 300 kB are coming from Ethers, RainbowKit and wagmi. So the app is not slow specifically because of RK but we need to make a collective effort to reduce the impact as much as possible.

A heavy bundle size means large network payload and more time spent parsing, compiling and executing JavaScript. This is annoying if the code is unused (which is probably the case for Ethers, I doubt that we need the whole library).

> i like your suggestions and would be open to welcome such a contribution! <3

Cool, glad to hear that!

> cc'ing @markdalgleish here in case this is something he's been thinking about already
 
Yeah I'd love to know if you already thought about this issue

**Reply by peduarte** [reply]
💯 

happy to start off by improving `ethers` tree-shaking. cc'ing @jxom here as well, in case this can also be done within `wagmi`

**Reply by jeevanpillay0000** [reply]
Some research I have done

There could be a big bundle size improve on the core rainbow-kit package by only using `@ethersproject/providers`

Looking at the dependencies in rainbow-kit, the `@ethersproject/providers` is already a core dependency, however, `ethers` was readded into the dependency and invoke a couple times in the project only to use the same functionality that `@ethersproject/providers` offer (see image below for where `ethers` is invoked)

<img width="303" alt="Screen Shot 2022-09-11 at 4 04 57 pm" src="https://user-images.githubusercontent.com/33827489/189514753-8c12557b-65ff-4bde-b01a-7725e83849b0.png">

According to bundlephobia, the `@ethersproject/providers`  has a **gzipped size of 81kB** while the `ethers` package is whopping **gzipped size of 133.2kB**

Unfortunately, `wagmi` also has haphazardly uses the `ethers` package as a whole. Hoping we can get them to change as well and improve everyones load speeds.

**Comment by DanielSinclair** [comment]
RainbowKit 1.0 and beyond should do much better here as the `wagmi` team made great tree-shaking progress with `viem` and RainbowKit no longer directly relies on `ethers`
